### PR TITLE
Add UI to change Sacct command line

### PR DESCRIPTION
**Pull request by hv15**
This PR is in response to #15.

In the PR:
- add support for `-S`/starttime and `-E`/endtime entry for `sacct` command
- change config file to use above command line values (`-d` and `-t` flags have been removed) 
- add UI text showing above values
- add UI inputboxes for the above commands (user can make change directly in JobHist panel by pressing the `t` key)

Extras (might move out to other PRs):
- add refresh options by `r` key (currently only programmed for JobHist panel, can be extended)
- ~add lipgloss styling to filter and time range values (not very well done, I don't have a very good sense of UI aesthetics)~

Todos:
- [ ] ~resolve user `sacct` accounts inconsistency, see below for info (1).~ (for another PR)

Further details (1):
`sacct` and `sacctmgr` have a somewhat subtle difference in behavior if they are called on a multi-cluster SLURM setup. By default the `sacct` only queries the _local_ SLURM cluster for information, whereas `sacctmgr` queries (and modifies) _all_ SLURM clusters. This means that when querying for user associated accounts, we get a list of accounts for all clusters, but calls to `sacct` with these might not match the _local_ cluster.

### CODE REVIEW

**Comment by pja237** [comment]
Hey @hv15 ,
just to let you know i'm back online after the move. Checked out this PR and it looks great so far!
Love the use of the multiple textinputs to manipulate sacct line :rocket: 

One small item to keep in mind while developing this feature, all new things shouldn't break the set minimum Width*Height:
https://github.com/CLIP-HPC/SlurmCommander/blob/21cee5ddc47eaad02dbdc37809f38085e194e6bf/internal/model/update.go#L344-L345
so while you're developing things that add lines/columns, be sure to turn on debug and test-resize your terminal to fit the minimum.
Can you just clarify for me the sacct/sacctmgs inconsistency? Not sure i understand?

Besides that, if i can be of any assistance on this, feel free to let me know, otherwise, amazing job!

**Comment by hv15** [comment]
@pja237  I've made sure that the height constraints aren't overstepped. The PR is 99% complete, the last bit is this issue with `sacct`/`sacctmgr` commands, which perhaps is not so much of an issue.

So on our cluster we have several SLURM clusters configured (different systems for different teams) that share the same slurmdb and job accounting. On our setup each SLURM cluster uses the same account associations (mostly), meaning that with the call of `sacctmgr` in scom to list out all the associate accounts we get duplicate accounts and/or accounts unique to a particular SLURM cluster. Now `sacct` is a very forgiving command, it neither checks the correctness of some parameters *nor* does it error out. This means that in most cases, sending a complete list of account associations for all SLURM clusters, even if we're only querying one of these clusters, should still work as we use the same account names between clusters (mostly).

The potential issue is if a user has a unique account association with one of the cluster, and does a query, `sacct` will return nothing unless we explicit query the cluster where the account association actually exists. There is no error or return code for this as far as I know, so we can't even check for this directly.

I think there are two ways to go about fixing this:
1. ask the user to select (or configure) the SLURM cluster(s) they want to interact with, and propagate this setting to all SLURM commands (using env `SLURM_CLUSTERS` for example).
2. do a complete query, setting the `-L` or `-a` flag for SLURM commands, so that we query all clusters (might cause permission errors)

Given the scope, maybe it makes sense to work on this in a separate PR?

File: internal/model/tabs/jobhisttab/jobhisttabview.go [hunk 1]
```diff
@@ -95,36 +95,50 @@ func (jh *JobHistTab) View(l *log.Logger) string {
 		MainWindow strings.Builder
 	)
 
+	// Show parameters to the user
+	heading := lipgloss.NewStyle().Bold(true).Foreground(lipgloss.Color("#FAFAFA")).Background(lipgloss.Color("#37ABBF")).Width(12).PaddingLeft(1).PaddingRight(1).Align(lipgloss.Right)
```
Comment (pja237) [inline internal/model/tabs/jobhisttab/jobhisttabview.go:99]: Can we keep the fg `styles.Yellow` and bg `styles.Blue` as color scheme?
Usually wouldn't care, but the whole coloring scheme was picked on purpose for this one.

File: internal/model/tabs/jobhisttab/jobhisttab.go [hunk 2]
```diff
@@ -46,9 +50,57 @@ type Breakdowns struct {
...
+	var tmp_t textinput.Model
+
+	// User Input (for modifying the table/view)
+	userinput := generic.UserInputs {
+		FocusIndex: 0,
+		ParamTexts: make([]string, 3),
+		Params: make([]textinput.Model, 3),
+	}
+
+	// TODO: we should probably think of encapsulating this
```
Comment (pja237) [inline internal/model/tabs/jobhisttab/jobhisttab.go:64]: Why not just do a package struct var of type generic.UserInputs in this package, then from scom.go just fill out these three fields that need changing and initiate the model with that?

**Review by pja237 — CHANGES_REQUESTED** [review]
Some bugs that i've noticed appeared with these changes:

---
On all tabs that support filter:
1. turn on filter, `/`, type anything `asdf`, apply
2. turn on filter, `ESC` should clear it, doesn't work

---

1. `c` to open counters
3. `r` to refresh
4. `Waiting for job history...` line is never cleared.

---

**Comment by pja237** [comment]
Hey @hv15,
that was a massive bit of work you've pulled there, respect!
Agree with everything, few minor wishes from my side and a few bugs creeped in, it's all written above.
If i haven't been precise about something, just let me know and i'll clarify myself.

Besides that, great work! :+1:

File: internal/keybindings/keybindings.go [hunk 3]
```diff
@@ -61,6 +64,18 @@ var DefaultKeyMap = KeyMap{
...
+		key.WithKeys("r"),
+		key.WithHelp("r", "Refresh View"),
+	),
+	TimeRange: key.NewBinding(
+		key.WithKeys("t"),
+		key.WithHelp("t", "Modify Time Range"),
+	),
+	Params: key.NewBinding(
+		key.WithKeys("P"),
+		key.WithHelp("P", "Modify Parameters"),
```
[the remark this answers is not in the document]
Comment (hv15) [reply internal/keybindings/keybindings.go]: agreed, removed.

File: internal/model/tabs/jobhisttab/jobhisttabview.go (hunk 1 above)
Comment (hv15) [reply internal/model/tabs/jobhisttab/jobhisttabview.go]: I only now just reads through styles.go, I've gone for the TextYellowOnBlue style.

File: internal/model/update.go [hunk 4]
```diff
@@ -631,6 +735,20 @@ func (m Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 				}
 			}
 
+		// Refresh the View
+		case key.Matches(msg, keybindings.DefaultKeyMap.Refresh):
+			switch m.ActiveTab {
+			case tabJobHist:
```
[the remark this answers is not in the document]
Comment (hv15) [reply internal/model/update.go:748]: done, also changed how keys are activated and deactivated, the latter is done explicitly now. Maybe this can be done better...

**Comment by hv15** [comment]
Hi, was AFK for the last few days but am back now; I've pushed some new commits which
- fixed `Waiting for job history...` line, this was a UI bug due to me adding an extra line without reducing the table height
- fixed filter bug, ooops I had deleted the line when I was still learning the code
- reworked key activation/deactivation, we now explicitly enable/disable keys which resolves issues like having `t` or `r` appearing in other tabs, or an issue that I ran into on doing `shift-tab` and then `tab` causing keys to mysteriously activate or deactivate without reason.
- minor UI change to the help window (footer) to have it extend to the entire width of the shell window (could do the same for the main window, but I think one needs to carefully consider the column content).

File: internal/model/tabs/jobhisttab/jobhisttab.go (hunk 2 above)
Comment (hv15) [reply internal/model/tabs/jobhisttab/jobhisttab.go:64]: My idea was that UserInputs should be _generic_ and used for defining a UI for any number of textinputs, really we could use this for the filter as well?

File: internal/model/update.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pja237) [reply internal/model/update.go:749]: This was on my mind for a while, __love__ this change :heart:

File: internal/model/tabs/jobhisttab/jobhisttab.go (hunk 2 above)
Comment (pja237) [reply internal/model/tabs/jobhisttab/jobhisttab.go:64]: Ah ok, understand now. Can be, needs a bit of thoughts and love before but yes, i agree with the idea. :+1:

**Comment by pja237** [comment]
I've given it a quick spin now and it looks very much polished and ready to roll-out.
Just let me know if it's all finished from your side before i release.

For the future, we should break up the work in smaller chunks, to keep the PRs from growing. 
I know there is so much impromptu code inside, and lots of dead ends and anti-patterns and whatnots. 
It's been a learning experience also for me using this architecture/framework so i was never quite sure how to best organize the code/data (still not sure in many cases :laughing: )
Knowing myself how tempting it is to stray from the path and start fixing legacy code, because so much could be refactored,  we should do it in smaller increments. Take note of something bad, make issue, fix in small increments i'd propose.

Perhaps keep a discussion thread/issue list/project tasks with the items that should/could be refactored?

**Review by pja237 — APPROVED** [review]

**Comment by hv15** [comment]
Hi @pja237 yes its done now.

I agree, this PR got way to big :rofl:  I'll open up an issue about some of the ideas we discussed, maybe these will become PR :smile_cat:
